colibre/scripts/bh_luminosities.py

```diff
@@ -0,0 +1,172 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import scipy.stats as stats
+
+from swiftsimio import load
+
+import unyt
+from unyt import mh, cm, Gyr, speed_of_light
+from matplotlib.colors import LogNorm
+from matplotlib.animation import FuncAnimation
+from matplotlib import rcParams
+
+# Set the limits of the figure.
+mass_bounds = [1e5, 3e10]
+value_bounds = [1e38, 1e48]
+bins = 25
+def_value = -1.0
+plot_scatter = True
+
+
+def get_data(filename):
+
+    data = load(filename)
+
+    masses = data.black_holes.subgrid_masses.to("Msun")
+
+    try:
+        rad_effs = data.black_holes.radiative_efficiencies
+        accr_rates = data.black_holes.accretion_rates.astype(np.float64).to(
+            unyt.kg / unyt.s
+        )
+        values = rad_effs * (speed_of_light ** 2) * accr_rates
+        values = values.to(unyt.erg / unyt.s)
+
+        # Make sure to take only black holes with a non-zero luminosity (efficiency)
+        masses = masses[rad_effs > 1e-6]
+        values = values[rad_effs > 1e-6]
+    except:
+        values = unyt.unyt_array(
+            np.zeros(masses.shape), dtype=np.float64, units=unyt.erg / unyt.s
+        )
+
+    return masses, values
+
+
+def calculate_medians(filename, mass_bounds, value_bounds, bins):
+
+    masses, values = get_data(filename)
+
+    masses_3rd_most_massive = np.sort(masses)[-3]
+
+    mass_bins = np.logspace(np.log10(mass_bounds[0]), np.log10(mass_bounds[1]), bins)
+    bin_width = (np.log10(mass_bounds[1]) - np.log10(mass_bounds[0])) / bins
+
+    threshold_mass = 10 ** (
+        np.log10(mass_bins[mass_bins < masses_3rd_most_massive][-1]) + bin_width * 0.5
+    )
+    mass_bins = mass_bins[mass_bins < threshold_mass]
+
+    values_most_massive = values[masses > threshold_mass]
+    masses_most_massive = masses[masses > threshold_mass]
+
+    values_rest = values[masses < threshold_mass]
+    masses_rest = masses[masses < threshold_mass]
+
+    medians, _, _ = stats.binned_statistic(
+        masses_rest, values_rest, statistic="median", bins=mass_bins
+    )
+    percentile_10s, _, _ = stats.binned_statistic(
+        masses_rest,
+        values_rest,
+        statistic=lambda x: np.percentile(x, 10.0),
+        bins=mass_bins,
+    )
+    percentile_90s, _, _ = stats.binned_statistic(
+        masses_rest,
+        values_rest,
+        statistic=lambda x: np.percentile(x, 90.0),
+        bins=mass_bins,
+    )
+    mass_bins = np.array(
+        [(mass_bins[i] + mass_bins[i + 1]) / 2 for i in range(np.size(medians))]
+    )
```

Comment:
I would probably call `binned_median_line() `from https://github.com/SWIFTSIM/velociraptor-python/blob/master/velociraptor/tools/lines.py#L165 instead of computing all these quantities myself. This would significantly shorten the code and thus improve its readability.

Reply:
Including this now in all of the scripts.

Reply:
Thanks!